### ISSUE TITLE
Scheduler does not evict engine from scheduling decision before it is flagged as dead

### ISSUE THREAD

**Issue by dongluochen** [docker/swarm#1328]
In Swarm 1.0.0 it may take 90 seconds or more to flag a node down. In a swarm set up with 2 nodes 'vm2' and 'vm3'. When docker daemon on vm2 was shutdown, 'docker run’ commands to swarm failed for around 90 seconds until swarm ‘flag engine as dead’. During that time requests to swarm got refused. In these 2 node setup it has about 50% chance to report the issue. 

```
dchen@vm3:~/go/src/github.com/docker/swarm$ docker -H 0.0.0.0:2375 info
Containers: 37
Images: 7
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 2
 vm2: 192.168.56.202:4444
  └ Containers: 18
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 4.053 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.19.0-25-generic, operatingsystem=Ubuntu 14.04.3 LTS, storagedriver=aufs
 vm3: 192.168.56.203:4444
  └ Containers: 19
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 4.053 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.19.0-25-generic, operatingsystem=Ubuntu 14.04.3 LTS, storagedriver=aufs
CPUs: 2
Total Memory: 8.106 GiB
Name: vm3
```

```
dchen@vm3:~/go/src/github.com/docker/swarm$ ./swarm manage 192.168.56.202:4444,192.168.56.203:4444
INFO[0000] Listening for HTTP                            addr=127.0.0.1:2375 proto=tcp
INFO[0000] Registered Engine vm3 at 192.168.56.203:4444
INFO[0000] Registered Engine vm2 at 192.168.56.202:4444
ERRO[3135] HTTP error: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused  status=500
ERRO[3146] HTTP error: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused  status=500
ERRO[3148] HTTP error: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused  status=500
ERRO[3151] HTTP error: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused  status=500
ERRO[3153] HTTP error: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused  status=500
ERRO[3231] Flagging engine as dead. Updated state failed 3 times: Get http://192.168.56.202:4444/v1.15/containers/json?all=1&size=0: dial tcp 192.168.56.202:4444: getsockopt: connection refused  id=S5HB:3CIV:DUKT:N2AP:UTA2:FNAZ:SYOJ:MXWY:UJCJ:D5YC:G3HJ:7TKO name=vm2
ERRO[3271] Flagging engine as dead. Updated state failed 4 times: Get http://192.168.56.202:4444/v1.15/containers/json?all=1&
```

```
dchen@vm3:~/go/src/github.com/docker/swarm$ sudo docker -H 0.0.0.0:2375 run hello-world
Error response from daemon: Post http://192.168.56.202:4444/v1.15/containers/create: dial tcp 192.168.56.202:4444: getsockopt: connection refused
```

```
dchen@vm3:~/go/src/github.com/docker/swarm$ docker -H 0.0.0.0:2375 version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 05:37:18 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      swarm/1.0.0
 API version:  1.16
 Go version:   go1.5.1
 Git commit:   HEAD
 Built:
 OS/Arch:      linux/amd64
```

**Comment by vieux** [comment]
ping @abronan @aluzzardi

**Comment by abronan** [comment]
The problem as I see it is that the node is not flagged as dead as soon as a connection fails. Thus in a setup of two nodes, the scheduler will continually pick `node2` even though it's down but still not flagged as dead. The result is that all the requests are going to fail for `90` seconds.

The fact that there are only two nodes exacerbate the problem.

Se we should evict the node from any scheduling decision as soon as we see a connection failure to an Engine.

**Comment by abronan** [comment]
There are two solutions:

- Either we say it's the default behavior to wait for the node being flagged as dead during the `refresh` loop. In this case it's only a problem when you have a small number of nodes.

Downside (let's say bug even): *You have N/M chances to be blocked for the time of the refresh loop if the node that is down is the one picked by the strategy for the next request*

- Either we flag as `unhealthy` as soon as a Create request of some sort (Container, Network, whatever using the scheduling) fails. In this case the incriminated node is not accessible and marked as `unhealthy` until the next refresh loop. 

Downside: *The network can behave erratically and if failing often the node will be virtually disabled but still listed*.

The fix is super straightforward (tested the second behavior and I think this is more acceptable than the first one).

Let me know what you think @docker/swarm-maintainers :cocktail:

**Comment by aluzzardi** [comment]
@dongluochen @abronan @vieux: That's the work I started with #1276, basically, instead of selecting one node to run the container, we get a list of nodes.

They idea is that we can try on one of the nodes (the first one, since they are ranked by priority) and if it doesn't work, we move on to the second one.

Also related to this, which we should probably push by `1.0` is: the number of retries as well as the duration should be configurable.

**Comment by abronan** [comment]
Oh forgot about this change. Indeed sounds like we just need to continue on that path and select the next one on the list :+1:

**Comment by dongluochen** [comment]
#1276 sounds great! If we have a short list of nodes, swarm manager can also 'ping' them to select healthy ones before containers are scheduled. It may reduce failure attempts to provide a better user experience. One failure case for that is swarm manager itself is under network partition that it cannot reach the majority of nodes. That problem should be handled by election/monitoring.

**Comment by dongluochen** [comment]
Closed by #1600.
